test/integration/scheduler_perf/create.go

```diff
@@ -19,71 +19,108 @@ package benchmark
+type createAny struct {
+	// Must match createAnyOpcode.
 	Opcode operationCode
 	// Namespace the object should be created in. Must be empty for cluster-scoped objects.
 	Namespace string
 	// Path to spec file describing the object to create.
 	TemplatePath string
 }
 
-func (cro *createOp[T, P]) isValid(allowParameterization bool) error {
-	var p P
-	if cro.Opcode != p.Opcode() {
-		return fmt.Errorf("invalid opcode %q; expected %q", cro.Opcode, p.Opcode())
-	}
-	if p.Namespaced() && cro.Namespace == "" {
-		return fmt.Errorf("Namespace must be set")
-	}
-	if !p.Namespaced() && cro.Namespace != "" {
-		return fmt.Errorf("Namespace must not be set")
+var _ runnableOp = &createAny{}
+
+func (c *createAny) isValid(allowParameterization bool) error {
+	if c.Opcode != createAnyOpcode {
+		return fmt.Errorf("invalid opcode %q; expected %q", c.Opcode, createAnyOpcode)
 	}
-	if cro.TemplatePath == "" {
+	if c.TemplatePath == "" {
 		return fmt.Errorf("TemplatePath must be set")
 	}
+	// The namespace can only be checked during later because we don't know yet
+	// whether the object is namespaced or cluster-scoped.
 	return nil
 }
 
-func (cro *createOp[T, P]) collectsMetrics() bool {
+func (c *createAny) collectsMetrics() bool {
 	return false
 }
 
-func (cro *createOp[T, P]) patchParams(w *workload) (realOp, error) {
-	return cro, cro.isValid(false)
+func (c *createAny) patchParams(w *workload) (realOp, error) {
+	return c, c.isValid(false)
 }
 
-func (cro *createOp[T, P]) requiredNamespaces() []string {
-	if cro.Namespace == "" {
+func (c *createAny) requiredNamespaces() []string {
+	if c.Namespace == "" {
 		return nil
 	}
-	return []string{cro.Namespace}
+	return []string{c.Namespace}
 }
 
-func (cro *createOp[T, P]) run(tCtx ktesting.TContext) {
-	var obj *T
-	var p P
-	if err := getSpecFromFile(&cro.TemplatePath, &obj); err != nil {
-		tCtx.Fatalf("parsing %s %q: %v", p.Name(), cro.TemplatePath, err)
+func (c *createAny) run(tCtx ktesting.TContext) {
+	var obj *unstructured.Unstructured
+	if err := getSpecFromFile(&c.TemplatePath, &obj); err != nil {
+		tCtx.Fatalf("%s: parsing failed: %v", c.TemplatePath, err)
 	}
-	if _, err := p.CreateCall(tCtx.Client(), cro.Namespace)(tCtx, obj, metav1.CreateOptions{}); err != nil {
-		tCtx.Fatalf("create %s: %v", p.Name(), err)
+
+	// Not caching the discovery result isn't very efficient, but good enough when
+	// createAny isn't done often.
+	discoveryCache := memory.NewMemCacheClient(tCtx.Client().Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryCache)
+	gv, err := schema.ParseGroupVersion(obj.GetAPIVersion())
+	if err != nil {
+		tCtx.Fatalf("%s: extract group+version from object %q: %v", c.TemplatePath, klog.KObj(obj), err)
 	}
-}
+	gk := schema.GroupKind{Group: gv.Group, Kind: obj.GetKind()}
+
+	create := func() error {
+		mapping, err := restMapper.RESTMapping(gk, gv.Version)
+		if err != nil {
+			// Cached mapping might be stale, refresh on next try.
```

Comment:
We have to call `restMapper.Reset()` before returning. Otherwise `restMapper` will never refresh itself.

Reply:
Note the `!d.cl.Fresh()`: this depends on the `discovery.CachedDiscoveryInterface` implementation detecting "stale" data.

If I remember correctly, it doesn't do that just because some result is not found and the `Reset` in the code above was needed.